File: ABC303/c.py
n,m,h,k = map(int, input().split())
s = input()
items = {tuple(map(int, input().split())) for _ in range(m)}

pos = [0,0]

for i in s:
    if i == 'R':
        pos[0] += 1
    elif i == 'L':
        pos[0] -=1
    elif i == 'U':
        pos[1] += 1
    else:
        pos[1] -=1
    h -= 1
    if h  == -1:
        break
    elif h<k and tuple(pos) in items :
        items.remove(tuple(pos))
        h = k

if(h >= 0):
    print('Yes')
else:
    print('No')

File: ABC303/c.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>

struct Position {
    int x, y;
};

int main() {
    int n, m, h, k;
    std::cin >> n >> m >> h >> k;

    std::vector<Position> items(m);
    for (int i = 0; i < m; i++) {
        std::cin >> items[i].x >> items[i].y;
    }

    Position pos = {0, 0};

    std::string s;
    std::cin >> s;

    for (char i : s) {
        if (i == 'R') {
            pos.x++;
        } else if (i == 'L') {
            pos.x--;
        } else if (i == 'U') {
            pos.y++;
        } else if (i == 'D') {
            pos.y--;
        }
        h--;
        if (h == -1) {
            break;
        }

        if (h < k) {
            auto it = std::find_if(items.begin(), items.end(), [&pos](const Position& item) {
                return item.x == pos.x && item.y == pos.y;
            });

            if (it != items.end()) {
                items.erase(it);
                h = k;
            }
        }
    }

    if (h >= 0) {
        std::cout << "Yes" << std::endl;
    } else {
        std::cout << "No" << std::endl;
    }

    return 0;
}
